rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,139 @@
+
+A single version of the Teleport client tools will be stored in the user's $HOME directory. Teleport already makes use of a .tsh directory for storing tsh config. Cached versions of tsh/tctl will also live within this directory under ` $HOME/.tsh/bin/{tsh,tctl}`. Whenever a new version of the client tools are available, the existing cached version will be replaced.
+
+The .tsh directory is only accessilbe to the user and we should make sure the new `bin` directory also only has user permissions.
+
+Everytime `login` or `status` command is executed, tsh will compare the currently cached version of the client tools with the current cloud-stable version of Teleport. If the cached version is behind or ahead of the current stable version, a new version of the client tools will be downloaded to replace the existing ones. When downloading the Teleport client tools, the binaries will be downloaded directly. We need to make sure to download a version that is compatible with the system.
+
+Because only a single version of the client tools will be cached at a time, we do not need to worry about cleaning up the cache.
+
+The client tools will need a way to execute the cached binaries. When tsh/tctl is executed, it will now need to decide whether to execute the command as usual, or fork a new process using the cached binary.
+
+One way to handle execution of cached binaries is to use the os.Executable function. os.Executable give us the path of the binary executable. The flow might look something like:
+- If executing `tsh login` execute as usual (cached tsh binary will not support TouchID on macOS)
+- If executable is in cache directory (.tsh/bin/tsh), then execute command as usual
+- Else try to execute tsh from cache if exists
+
+Another option could be to have users update their `PATH` env with `export PATH=$HOME/.tsh/bin:$PATH`. Adding the tsh bin directory to the `PATH` environment in the user's `.profile` can be handled by tsh. Users would then execute the cached binary directly. This seems like a simpler approach than the above methods. The problem is that the cached binary would not support TouchID for macOS users.
+
+I believe macOS software needs to be notarized as a requirement for TouchID. For this reason Teleport provides a separate signed tsh download at `https://cdn.teleport.dev/tsh-13.3.4.pkg`. We could change the download method for macOS users, but this might not be very straight forward to implement. The macOS `installer` would also require sudo/admin privileges to install the package. It sounds like this issue is being worked on currently, and we won't be shipping unsigned binaries after this quarter.
+
+### Config
+
+Automatic updates for client tools can be configured through tsh configuration. If `DISABLE_AUTO_UPDATE=true`, then auto updates for client tools will be disabled. If `DISABLE_UPDATE_PROMPT=true`, then the user will not be prompted to confirm the update. Both values will default to false.
```

Comment:
Are `DISABLE_AUTO_UPDATE` and `DISABLE_UPDATE_PROMPT` environment variables? It's not super clear what "tsh configuration" means in this case. If these are env. vars, let's say so explicitly.

I think it would also make sense to add these options to [tsh.yaml config](https://goteleport.com/docs/connect-your-client/tsh/#tsh-configuration-files) and give an example of what tsh config will look like.